corehq/blobs/s3db.py

```diff
@@ -0,0 +1,162 @@
+from __future__ import absolute_import
+import re
+from contextlib import contextmanager
+from uuid import uuid4
+
+from corehq.blobs import BlobInfo
+from corehq.blobs.exceptions import BadName, NotFound
+
+import boto3
+from botocore.handlers import calculate_md5
+from botocore.exceptions import ClientError
+from botocore.utils import fix_s3_host
+
+DEFAULT_S3_BUCKET = "blobdb"
+DEFAULT_BUCKET = "_default"
+SAFENAME = re.compile("^[a-z0-9_./-]+$", re.IGNORECASE)
+
+
+class S3BlobDB(object):
+
+    def __init__(self, settings):
+        self.db = boto3.resource(
+            's3',
+            endpoint_url=settings.get("url"),
+            aws_access_key_id=settings.get("access_key"),
+            aws_secret_access_key=settings.get("secret_key"),
+        )
+        self.s3_bucket_name = settings.get("s3_bucket", DEFAULT_S3_BUCKET)
+        self.s3_bucket_exists = False
+        # https://github.com/boto/boto3/issues/259
+        self.db.meta.client.meta.events.unregister('before-sign.s3', fix_s3_host)
+
+    def put(self, content, basename="", bucket=DEFAULT_BUCKET, content_md5=None):
```

Comment:
Yeah, our concept of a bucket is slightly different from the S3 concept of a bucket. All S3 bucket variables start with `s3_bucket` to hopefully mitigate any potential confusion.